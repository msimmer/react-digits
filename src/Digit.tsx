import React from 'react'
import styled from 'styled-components'
import 'styled-components/macro'
import PropTypes from 'prop-types'
import { theme } from './theme'
import { DigitProps, ThemeProps } from './types'
import { arrows } from './css'

const Number = styled.div<ThemeProps>`
  *,
  *::before,
  *::after {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
  }

  box-sizing: border-box;
  margin: 0;
  padding: 0;
  position: relative;
  width: ${(props: ThemeProps) => props.theme.width}px;
  display: flex;
  align-items: center;
  flex-direction: row;
  flex-wrap: wrap;

  div {
    position: relative;
    background-color: ${(props: ThemeProps) => props.theme.lineColor};
  }

  &.zero {
    div:nth-of-type(4) {
      opacity: 0;
    }
  }
  &.one {
    div:nth-of-type(1),
    div:nth-of-type(2),
    div:nth-of-type(4),
    div:nth-of-type(5),
    div:nth-of-type(6) {
      opacity: 0;
    }
  }
  &.two {
    div:nth-of-type(1),
    div:nth-of-type(7) {
      opacity: 0;
    }
  }
  &.three {
    div:nth-of-type(1),
    div:nth-of-type(5) {
      opacity: 0;
    }
  }
  &.four {
    div:nth-of-type(2),
    div:nth-of-type(5),
    div:nth-of-type(6) {
      opacity: 0;
    }
  }
  &.five {
    div:nth-of-type(3),
    div:nth-of-type(5) {
      opacity: 0;
    }
  }
  &.six {
    div:nth-of-type(3) {
      opacity: 0;
    }
  }
  &.seven {
    div:nth-of-type(1),
    div:nth-of-type(4),
    div:nth-of-type(5),
    div:nth-of-type(6) {
      opacity: 0;
    }
  }
  &.eight {
  }
  &.nine {
    div:nth-of-type(5) {
      opacity: 0;
    }
  }
`

const Vertical = styled.div<ThemeProps>`
  ${arrows('vertical')}
  flex: 0 0 ${(props: ThemeProps) => props.theme.lineWidth}px;
`

const VerticalTop = styled(Vertical)<ThemeProps>`
  height: ${(props: ThemeProps) => props.theme.height}px;
  margin: 12px 0 2px;
`

const VerticalBottom = styled(Vertical)<ThemeProps>`
  height: ${(props: ThemeProps) => props.theme.height}px;
  margin: 2px 0 12px;
`

const Horizontal = styled.div<ThemeProps>`
  ${arrows('horizontal')}
  flex: 0 0 ${(props: ThemeProps) => props.theme.lineWidth}px;
`

const HorizontalMiddle = styled(Horizontal)<ThemeProps>`
  border: 0;
  background-color: ${(props: ThemeProps) => props.theme.lineColor};
  flex: 0 0 calc(100% - ${(props: ThemeProps) => props.theme.lineWidth * 2}px);
  height: ${(props: ThemeProps) => props.theme.lineWidth}px;
  margin: 0 ${(props: ThemeProps) => props.theme.lineWidth}px;
`

const HorizontalTop = styled(Horizontal)<ThemeProps>`
  border: 0;
  background-color: ${(props: ThemeProps) => props.theme.lineColor};
  flex: 1 1 calc(100% - ${(props: ThemeProps) => props.theme.lineWidth}px * 4);
  height: ${(props: ThemeProps) => props.theme.lineWidth}px;
  align-self: flex-start;
`

const HorizontalBottom = styled(HorizontalTop)<ThemeProps>`
  align-self: flex-end;
`

class Digit extends React.Component<DigitProps, {}> {
  static defaultProps = { className: 'eight', ...theme }

  // show runtime errors in dev mode
  static propTypes = {
    className: PropTypes.oneOf(['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine', 'zero']),
  }

  render() {
    return (
      <Number className={this.props.className}>
        <VerticalTop />
        <HorizontalTop />
        <VerticalTop />
        <HorizontalMiddle />
        <VerticalBottom />
        <HorizontalBottom />
        <VerticalBottom />
      </Number>
    )
  }
}

export default Digit
