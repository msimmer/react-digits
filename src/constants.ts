import { BordersIndex, BordersValue, OrientationsIndex, OrientationsValue, DigitProp } from './types'

export const borders: { readonly [k in BordersIndex]: BordersValue } = {
  TOP: 'top',
  RIGHT: 'right',
  BOTTOM: 'bottom',
  LEFT: 'left',
}

export const orientations: { readonly [k in OrientationsIndex]: OrientationsValue } = {
  VERTICAL: 'vertical',
  HORIZONTAL: 'horizontal',
}

// prettier-ignore
// export const digits: { readonly [k in DigitProp]: number[] } = {
export const digits: { readonly [key: string]: number[] } = {
  zero:  [1, 1, 1, 0, 1, 1, 1],
  one:   [0, 0, 1, 0, 0, 0, 1],
  two:   [0, 1, 1, 1, 1, 1, 0],
  three: [0, 1, 1, 1, 0, 1, 1],
  four:  [1, 0, 1, 1, 0, 0, 1],
  five:  [1, 1, 0, 1, 0, 1, 1],
  six:   [1, 1, 0, 1, 1, 1, 1],
  seven: [0, 1, 1, 0, 0, 0, 1],
  eight: [1, 1, 1, 1, 1, 1, 1],
  nine:  [1, 1, 1, 1, 0, 1, 1],
}
