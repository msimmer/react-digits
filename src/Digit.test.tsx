import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import { configure, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import Digit from './Digit'
import { digits } from './constants'

configure({ adapter: new Adapter() })

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<Digit />, div)
  ReactDOM.unmountComponentAtNode(div)
})

it('matches the snapshot', () => {
  const tree = renderer.create(<Digit />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders the default digit', () => {
  const wrapper = mount(<Digit />)
  expect(wrapper.find('TopLeft')).toHaveStyleRule('opacity', '1')
  expect(wrapper.find('TopMiddle')).toHaveStyleRule('opacity', '1')
  expect(wrapper.find('TopRight')).toHaveStyleRule('opacity', '1')
  expect(wrapper.find('Middle')).toHaveStyleRule('opacity', '1')
  expect(wrapper.find('BottomLeft')).toHaveStyleRule('opacity', '1')
  expect(wrapper.find('BottomMiddle')).toHaveStyleRule('opacity', '1')
  expect(wrapper.find('BottomRight')).toHaveStyleRule('opacity', '1')
})

it('renders the digits', () => {
  const wrapper = mount(<Digit />)
  const numbers = ['zero', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine']
  const components = ['TopLeft', 'TopMiddle', 'TopRight', 'Middle', 'BottomLeft', 'BottomMiddle', 'BottomRight']
  numbers.forEach(digit => {
    wrapper.setProps({ digit })
    digits[digit].forEach((opacity: number, i: number) => {
      expect(wrapper.find(components[i])).toHaveStyleRule('opacity', String(opacity))
    })
  })
})

it('accepts props', () => {
  const wrapper = mount(<Digit width={10} height={10} lineColor={'green'} lineWidth={10} />)
  expect(wrapper.childAt(0)).toHaveStyleRule('width', '10px')
  expect(wrapper.find('TopLeft')).toHaveStyleRule('height', '10px')
  expect(wrapper.find('TopLeft')).toHaveStyleRule('background-color', 'green')
  expect(wrapper.find('TopLeft')).toHaveStyleRule('flex', '0 0 10px')
})
