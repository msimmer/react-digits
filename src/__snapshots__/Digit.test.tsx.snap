// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  position: relative;
  width: 70px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 *,
.c0 *::before,
.c0 *::after {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

.c1 {
  position: relative;
  -webkit-flex: 0 0 10px;
  -ms-flex: 0 0 10px;
  flex: 0 0 10px;
  background-color: red;
  position: relative;
  margin: 12px 0 2px;
  height: 40px;
  background-color: red;
  opacity: 1;
}

.c1::before {
  content: '';
  display: block;
  position: absolute;
  height: 0;
  width: 0;
  background-color: transparent;
  border-top: 5px solid transparent;
  border-right: 5px solid transparent;
  border-left: 5px solid transparent;
  border-bottom: 5px solid transparent;
  border-bottom: 5px solid red;
  left: 0;
  top: -10px;
}

.c1::after {
  content: '';
  display: block;
  position: absolute;
  height: 0;
  width: 0;
  background-color: transparent;
  border-top: 5px solid transparent;
  border-right: 5px solid transparent;
  border-left: 5px solid transparent;
  border-bottom: 5px solid transparent;
  border-top: 5px solid red;
  left: 0;
  bottom: -10px;
}

.c2 {
  position: relative;
  -webkit-flex: 0 0 10px;
  -ms-flex: 0 0 10px;
  flex: 0 0 10px;
  background-color: red;
  border: 0;
  position: relative;
  background-color: red;
  -webkit-flex: 1 1 calc(100% - 10px * 4);
  -ms-flex: 1 1 calc(100% - 10px * 4);
  flex: 1 1 calc(100% - 10px * 4);
  height: 10px;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  opacity: 1;
}

.c2::before {
  content: '';
  display: block;
  position: absolute;
  height: 0;
  width: 0;
  background-color: transparent;
  border-top: 5px solid transparent;
  border-right: 5px solid transparent;
  border-left: 5px solid transparent;
  border-bottom: 5px solid transparent;
  border-right: 5px solid red;
  top: 0;
  left: -10px;
}

.c2::after {
  content: '';
  display: block;
  position: absolute;
  height: 0;
  width: 0;
  background-color: transparent;
  border-top: 5px solid transparent;
  border-right: 5px solid transparent;
  border-left: 5px solid transparent;
  border-bottom: 5px solid transparent;
  border-left: 5px solid red;
  top: 0;
  right: -10px;
}

.c3 {
  position: relative;
  -webkit-flex: 0 0 10px;
  -ms-flex: 0 0 10px;
  flex: 0 0 10px;
  background-color: red;
  position: relative;
  margin: 12px 0 2px;
  height: 40px;
  background-color: red;
  opacity: 1;
}

.c3::before {
  content: '';
  display: block;
  position: absolute;
  height: 0;
  width: 0;
  background-color: transparent;
  border-top: 5px solid transparent;
  border-right: 5px solid transparent;
  border-left: 5px solid transparent;
  border-bottom: 5px solid transparent;
  border-bottom: 5px solid red;
  left: 0;
  top: -10px;
}

.c3::after {
  content: '';
  display: block;
  position: absolute;
  height: 0;
  width: 0;
  background-color: transparent;
  border-top: 5px solid transparent;
  border-right: 5px solid transparent;
  border-left: 5px solid transparent;
  border-bottom: 5px solid transparent;
  border-top: 5px solid red;
  left: 0;
  bottom: -10px;
}

.c4 {
  position: relative;
  -webkit-flex: 0 0 10px;
  -ms-flex: 0 0 10px;
  flex: 0 0 10px;
  background-color: red;
  border: 0;
  position: relative;
  background-color: red;
  -webkit-flex: 0 0 calc(100% - 20px);
  -ms-flex: 0 0 calc(100% - 20px);
  flex: 0 0 calc(100% - 20px);
  height: 10px;
  margin: 0 10px;
  opacity: 1;
}

.c4::before {
  content: '';
  display: block;
  position: absolute;
  height: 0;
  width: 0;
  background-color: transparent;
  border-top: 5px solid transparent;
  border-right: 5px solid transparent;
  border-left: 5px solid transparent;
  border-bottom: 5px solid transparent;
  border-right: 5px solid red;
  top: 0;
  left: -10px;
}

.c4::after {
  content: '';
  display: block;
  position: absolute;
  height: 0;
  width: 0;
  background-color: transparent;
  border-top: 5px solid transparent;
  border-right: 5px solid transparent;
  border-left: 5px solid transparent;
  border-bottom: 5px solid transparent;
  border-left: 5px solid red;
  top: 0;
  right: -10px;
}

.c5 {
  position: relative;
  -webkit-flex: 0 0 10px;
  -ms-flex: 0 0 10px;
  flex: 0 0 10px;
  background-color: red;
  height: 40px;
  position: relative;
  margin: 2px 0 12px;
  background-color: red;
  opacity: 1;
}

.c5::before {
  content: '';
  display: block;
  position: absolute;
  height: 0;
  width: 0;
  background-color: transparent;
  border-top: 5px solid transparent;
  border-right: 5px solid transparent;
  border-left: 5px solid transparent;
  border-bottom: 5px solid transparent;
  border-bottom: 5px solid red;
  left: 0;
  top: -10px;
}

.c5::after {
  content: '';
  display: block;
  position: absolute;
  height: 0;
  width: 0;
  background-color: transparent;
  border-top: 5px solid transparent;
  border-right: 5px solid transparent;
  border-left: 5px solid transparent;
  border-bottom: 5px solid transparent;
  border-top: 5px solid red;
  left: 0;
  bottom: -10px;
}

.c6 {
  position: relative;
  -webkit-flex: 0 0 10px;
  -ms-flex: 0 0 10px;
  flex: 0 0 10px;
  background-color: red;
  border: 0;
  position: relative;
  background-color: red;
  -webkit-flex: 1 1 calc(100% - 10px * 4);
  -ms-flex: 1 1 calc(100% - 10px * 4);
  flex: 1 1 calc(100% - 10px * 4);
  height: 10px;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
  opacity: 1;
}

.c6::before {
  content: '';
  display: block;
  position: absolute;
  height: 0;
  width: 0;
  background-color: transparent;
  border-top: 5px solid transparent;
  border-right: 5px solid transparent;
  border-left: 5px solid transparent;
  border-bottom: 5px solid transparent;
  border-right: 5px solid red;
  top: 0;
  left: -10px;
}

.c6::after {
  content: '';
  display: block;
  position: absolute;
  height: 0;
  width: 0;
  background-color: transparent;
  border-top: 5px solid transparent;
  border-right: 5px solid transparent;
  border-left: 5px solid transparent;
  border-bottom: 5px solid transparent;
  border-left: 5px solid red;
  top: 0;
  right: -10px;
}

.c7 {
  position: relative;
  -webkit-flex: 0 0 10px;
  -ms-flex: 0 0 10px;
  flex: 0 0 10px;
  background-color: red;
  height: 40px;
  position: relative;
  margin: 2px 0 12px;
  background-color: red;
  opacity: 1;
}

.c7::before {
  content: '';
  display: block;
  position: absolute;
  height: 0;
  width: 0;
  background-color: transparent;
  border-top: 5px solid transparent;
  border-right: 5px solid transparent;
  border-left: 5px solid transparent;
  border-bottom: 5px solid transparent;
  border-bottom: 5px solid red;
  left: 0;
  top: -10px;
}

.c7::after {
  content: '';
  display: block;
  position: absolute;
  height: 0;
  width: 0;
  background-color: transparent;
  border-top: 5px solid transparent;
  border-right: 5px solid transparent;
  border-left: 5px solid transparent;
  border-bottom: 5px solid transparent;
  border-top: 5px solid red;
  left: 0;
  bottom: -10px;
}

<div
  className="c0"
>
  <div
    className="c1"
  />
  <div
    className="c2"
  />
  <div
    className="c3"
  />
  <div
    className="c4"
  />
  <div
    className="c5"
  />
  <div
    className="c6"
  />
  <div
    className="c7"
  />
</div>
`;
